jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c

```diff
@@ -406,26 +406,52 @@ ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this arg
                   ecma_op_to_string (this_arg),
                   ret_value);
 
-  ecma_value_t regexp_value = ECMA_VALUE_EMPTY;
+  *this_to_string_value_ptr = ecma_copy_value (this_to_string_value);
+
   /* 3. */
-  if (ecma_is_value_object (arg)
-      && ecma_object_class_is (ecma_get_object_from_value (arg), LIT_MAGIC_STRING_REGEXP_UL))
+  if (ecma_is_value_object (regexp_arg)
+      && ecma_object_class_is (ecma_get_object_from_value (regexp_arg), LIT_MAGIC_STRING_REGEXP_UL))
   {
-    regexp_value = ecma_copy_value (arg);
+    *regexp_value = ecma_copy_value (regexp_arg);
   }
   else
   {
     /* 4. */
-    ecma_value_t regexp_arguments[1] = { arg };
+    ecma_value_t regexp_arguments[1] = { regexp_arg };
     ECMA_TRY_CATCH (new_regexp_value,
                     ecma_builtin_regexp_dispatch_construct (regexp_arguments, 1),
                     ret_value);
 
-    regexp_value = ecma_copy_value (new_regexp_value);
+    *regexp_value = ecma_copy_value (new_regexp_value);
 
     ECMA_FINALIZE (new_regexp_value);
   }
 
+  ECMA_FINALIZE (this_to_string_value);
+  ECMA_FINALIZE (this_check_coercible_value);
+
+  return ret_value;
+} /* ecma_builtin_string_prepare_search */
+
+/**
+ * The String.prototype object's 'match' routine
+ *
+ * See also:
+ *          ECMA-262 v5, 15.5.4.10
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this argument */
+                                            ecma_value_t regexp_arg) /**< routine's argument */
+{
+  ecma_value_t this_to_string_value = ECMA_VALUE_EMPTY;
+  ecma_value_t regexp_value = ECMA_VALUE_EMPTY;
+
+  ecma_value_t ret_value = ecma_builtin_string_prepare_search (this_arg, &this_to_string_value,
+                                                               regexp_arg, &regexp_value);
```

Comment:
I think we can reduce the number of parameters of `ecma_builtin_string_prepare_search`. `this_to_string_value` can be returned by the function if no error happened.

Reply:
How should we handle the errors produced in `ecma_builtin_string_prepare_search`? Is there a way to decide whether the returned `this_to_string_value` contains an error or a valid value?

Reply:
Yes, you can use `ECMA_IS_VALUE_ERROR`

Reply:
Well I've started looking into this, but honestly this makes the code ugly and less readable to me. Personally I'd still prefer the current 2 input + 2 output + 1 retval setup.